engine/runtime/src/main/java/org/enso/interpreter/node/callable/thunk/ThunkExecutorNode.java

```diff
@@ -93,4 +90,50 @@ Stateful doUncached(
       }
     }
   }
+
+  static InvokeFunctionNode buildInvokeFunctionNode(BaseNode.TailStatus tailStatus) {
+    var node =
+        InvokeFunctionNode.build(
+            new CallArgumentInfo[0],
+            InvokeCallableNode.DefaultsExecutionMode.EXECUTE,
+            InvokeCallableNode.ArgumentsExecutionMode.EXECUTE);
+    node.setTailStatus(tailStatus);
+    return node;
+  }
+
+  @Specialization(
+      guards = {"!fn.isThunk()", "fn.isFullyApplied()", "isTail == cachedIsTail"},
+      limit = "3" /* limit is 3 because that's the number of different values for isTail */)
```

Comment:
One could define
```java
  static int numberOfDifferentValuesForIsTail() {
    return BaseNode.TailStatus.values().length;
  }
```
and then use
```suggestion
      limit = "numberOfDifferentValuesForIsTail()"
```